File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.js
function solution(enroll, referral, seller, amount) {
    const moneyMap = {};
    const refMap = {};
    for(let i = 0; i < enroll.length; i++){
        refMap[enroll[i]] = referral[i];
        moneyMap[enroll[i]] = 0;
    }
    for(let i = 0; i < seller.length; i++){
        let person = seller[i];
        let money = amount[i] * 100;
        while(person !== '-' && money > 0) {
            let distribute = Math.floor(money * 0.1);
            moneyMap[person] += money - distribute;
            money = distribute;
            person = refMap[person];
        }
    }
    return enroll.map(name => moneyMap[name]);
}

File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.py
import math

def solution(enroll, referral, seller, amount):
    ref = {}
    result = {}
    for i in range(len(enroll)):
        ref[enroll[i]] = referral[i]
        result[enroll[i]] = 0
    
    for i in range(len(seller)):
        p, cash = seller[i], amount[i] * 100
        while p != '-' and cash > 0:
            p_cash = math.ceil(cash * 0.9)
            result[p] += p_cash
            cash -= p_cash
            p = ref[p]
    
    answer = []
    for p in enroll:
        answer.append(result[p])
    return answer